### PR TITLE
feat(core): introduce support for animations

### PR DESCRIPTION
**Pull request by matsko**

### CODE REVIEW

**Comment by matsko** [comment]
This is what the codegen code looks like:

```
(function(jit_uninitialized0,jit_AnimationStyleUtil1,jit_AnimationSequencePlayer2,jit_AnimationStyles3,jit_AnimationKeyframe4,jit_StaticNodeDebugInfo5,jit_TemplateRef6,jit_NgFor7,jit_DebugAppView8,jit_ViewType_COMPONENT9,jit_ChangeDetectionStrategy_CheckAlways10,jit_AppElement11,jit_TemplateRef_12,jit_IterableDiffers13,jit_checkBinding14,jit_ViewEncapsulation_Emulated15,jit_val16,jit_interpolate17
/**/) {
function AnimateApp_boxAnimation_0(view,element,currentState,nextState) {
  var animationStateStyles = {
    'void': {
      'height': 0,
      'opacity': 0
    }
    ,
    'start': {
      'background': 'red',
      'height': '*'
    }
    ,
    'active': {
      'background': 'orange',
      'color': 'white',
      'font-size': '100px'
    }

  }
  ;
  var collectedStyles = {};
  var player = null;
  if ((currentState == jit_uninitialized0)) { (currentState = 'void'); }
  if ((nextState == jit_uninitialized0)) { (nextState = 'void'); }
  var startStateStyles = animationStateStyles[currentState];
  var finalStateStyles = animationStateStyles[nextState];
  view.renderer.setElementStyles(element,jit_AnimationStyleUtil1.clearStyles(startStateStyles));
  if ((((player == null) && (currentState == 'active')) && (nextState == 'start'))) { (player = new jit_AnimationSequencePlayer2([
    view.renderer.animate(element,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[startStateStyles])),[
      new jit_AnimationKeyframe4(0,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[{'transform': 'true'}]))),
      new jit_AnimationKeyframe4(1,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[{'transform': 'scale(2)'}])))
    ]
    ,500,0,null),
    view.renderer.animate(element,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[])),jit_AnimationStyleUtil1.balanceKeyframes(collectedStyles,finalStateStyles,[
      new jit_AnimationKeyframe4(0,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[{}]))),
      new jit_AnimationKeyframe4(1,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[{}])))
    ]
    ),500,0,null)
  ]
  )); }
  if ((player == null)) { (player = new jit_AnimationSequencePlayer2([
    view.renderer.animate(element,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[startStateStyles])),[
        new jit_AnimationKeyframe4(0,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[{
          'opacity': 'true',
          'height': 'true'
        }
      ]))),
        new jit_AnimationKeyframe4(1,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[{
          'opacity': 1,
          'height': 300
        }
      ])))
    ]
    ,1000,0,null),
    view.renderer.animate(element,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[])),[
      new jit_AnimationKeyframe4(0,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[{'background': 'true'}]))),
      new jit_AnimationKeyframe4(1,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[{'background': 'blue'}])))
    ]
    ,1000,0,null),
    view.renderer.animate(element,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[])),[
        new jit_AnimationKeyframe4(0,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[{
          'background': 'blue',
          'color': 'black'
        }
      ]))),
        new jit_AnimationKeyframe4(0.5,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[{
          'background': 'brown',
          'color': 'black'
        }
      ]))),
        new jit_AnimationKeyframe4(1,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[{
          'background': 'black',
          'color': 'white'
        }
      ])))
    ]
    ,1000,0,null),
    view.renderer.animate(element,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[])),jit_AnimationStyleUtil1.balanceKeyframes(collectedStyles,finalStateStyles,[
      new jit_AnimationKeyframe4(0,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[{}]))),
      new jit_AnimationKeyframe4(1,new jit_AnimationStyles3(jit_AnimationStyleUtil1.collectAndResolveStyles(collectedStyles,[{}])))
    ]
    ),2000,0,null)
  ]
  )); }
  player.onDone(function() {
    view.renderer.setElementStyles(element,jit_AnimationStyleUtil1.balanceStyles(startStateStyles,finalStateStyles));
  });
  view.registerActiveAnimation(player);
  player.play();
  return player;
}
var nodeDebugInfos_AnimateApp0 = [
  new jit_StaticNodeDebugInfo5([],null,{}),
  new jit_StaticNodeDebugInfo5([],null,{}),
  new jit_StaticNodeDebugInfo5([],null,{}),
  new jit_StaticNodeDebugInfo5([],null,{}),
  new jit_StaticNodeDebugInfo5([],null,{}),
  new jit_StaticNodeDebugInfo5([],null,{}),
  new jit_StaticNodeDebugInfo5([],null,{}),
  new jit_StaticNodeDebugInfo5([],null,{}),
  new jit_StaticNodeDebugInfo5([],null,{}),
  new jit_StaticNodeDebugInfo5([],null,{}),
  new jit_StaticNodeDebugInfo5([],null,{}),
  new jit_StaticNodeDebugInfo5([
    jit_TemplateRef6,
    jit_NgFor7
  ]
  ,null,{}),
  new jit_StaticNodeDebugInfo5([],null,{})
]
;
var renderType_AnimateApp = null;
function _View_AnimateApp0(viewUtils,parentInjector,declarationEl) {
  var self = this;
  jit_DebugAppView8.call(this, _View_AnimateApp0,renderType_AnimateApp,jit_ViewType_COMPONENT9,viewUtils,parentInjector,declarationEl,jit_ChangeDetectionStrategy_CheckAlways10,nodeDebugInfos_AnimateApp0);
}
_View_AnimateApp0.prototype = Object.create(jit_DebugAppView8.prototype);
_View_AnimateApp0.prototype.createInternal = function(rootSelector) {
  var self = this;
  var parentRenderNode = self.renderer.createViewRoot(self.declarationAppElement.nativeElement);
  self._el_0 = self.renderer.createElement(parentRenderNode,'button',self.debug(0,0,0));
  self._text_1 = self.renderer.createText(self._el_0,'Start State',self.debug(1,0,32));
  self._text_2 = self.renderer.createText(parentRenderNode,'\n',self.debug(2,0,52));
  self._el_3 = self.renderer.createElement(parentRenderNode,'button',self.debug(3,1,0));
  self._text_4 = self.renderer.createText(self._el_3,'Active State',self.debug(4,1,33));
  self._text_5 = self.renderer.createText(parentRenderNode,'\n',self.debug(5,1,54));
  self._el_6 = self.renderer.createElement(parentRenderNode,'button',self.debug(6,2,0));
  self._text_7 = self.renderer.createText(self._el_6,'Void State',self.debug(7,2,31));
  self._text_8 = self.renderer.createText(parentRenderNode,'\n',self.debug(8,2,50));
  self._el_9 = self.renderer.createElement(parentRenderNode,'hr',self.debug(9,3,0));
  self._text_10 = self.renderer.createText(parentRenderNode,'\n',self.debug(10,3,6));
  self._anchor_11 = self.renderer.createTemplateAnchor(parentRenderNode,self.debug(11,4,0));
  self._appEl_11 = new jit_AppElement11(11,null,self,self._anchor_11);
  self._TemplateRef_11_5 = new jit_TemplateRef_12(self._appEl_11,viewFactory_AnimateApp1);
  self._NgFor_11_6 = new jit_NgFor7(self._appEl_11.vcRef,self._TemplateRef_11_5,self.parentInjector.get(jit_IterableDiffers13),self.ref);
  self._text_12 = self.renderer.createText(parentRenderNode,'\n',self.debug(12,6,6));
  var disposable_0 = self.renderer.listen(self._el_0,'click',self.eventHandler(self._handle_click_0_0.bind(self)));
  var disposable_1 = self.renderer.listen(self._el_3,'click',self.eventHandler(self._handle_click_3_0.bind(self)));
  var disposable_2 = self.renderer.listen(self._el_6,'click',self.eventHandler(self._handle_click_6_0.bind(self)));
  self._expr_3 = jit_uninitialized0;
  self.init([],[
    self._el_0,
    self._text_1,
    self._text_2,
    self._el_3,
    self._text_4,
    self._text_5,
    self._el_6,
    self._text_7,
    self._text_8,
    self._el_9,
    self._text_10,
    self._anchor_11,
    self._text_12
  ]
  ,[
    disposable_0,
    disposable_1,
    disposable_2
  ]
  ,[]);
  return null;
};
_View_AnimateApp0.prototype.injectorGetInternal = function(token,requestNodeIndex,notFoundResult) {
  var self = this;
  if (((token === jit_TemplateRef6) && (11 === requestNodeIndex))) { return self._TemplateRef_11_5; }
  if (((token === jit_NgFor7) && (11 === requestNodeIndex))) { return self._NgFor_11_6; }
  return notFoundResult;
};
_View_AnimateApp0.prototype.detectChangesInternal = function(throwOnChange) {
  var self = this;
  self.debug(11,4,5);
  var currVal_3 = self.context.items;
  if (jit_checkBinding14(throwOnChange,self._expr_3,currVal_3)) {
    self._NgFor_11_6.ngForOf = currVal_3;
    self._expr_3 = currVal_3;
  }
  if (!throwOnChange) { self._NgFor_11_6.ngDoCheck(); }
  self.detectContentChildrenChanges(throwOnChange);
  self.detectViewChildrenChanges(throwOnChange);
};
_View_AnimateApp0.prototype._handle_click_0_0 = function($event) {
  var self = this;
  self.markPathToRootAsCheckOnce();
  self.debug(0,0,8);
  var pd_0 = ((self.context.state = 'start') !== false);
  return (true && pd_0);
};
_View_AnimateApp0.prototype._handle_click_3_0 = function($event) {
  var self = this;
  self.markPathToRootAsCheckOnce();
  self.debug(3,1,8);
  var pd_0 = ((self.context.state = 'active') !== false);
  return (true && pd_0);
};
_View_AnimateApp0.prototype._handle_click_6_0 = function($event) {
  var self = this;
  self.markPathToRootAsCheckOnce();
  self.debug(6,2,8);
  var pd_0 = ((self.context.state = 'void') !== false);
  return (true && pd_0);
};
function viewFactory_AnimateApp0(viewUtils,parentInjector,declarationEl) {
  if ((renderType_AnimateApp === null)) { (renderType_AnimateApp = viewUtils.createRenderComponentType('animate-app.html',0,jit_ViewEncapsulation_Emulated15,jit_val16)); }
  return new _View_AnimateApp0(viewUtils,parentInjector,declarationEl);
}
var nodeDebugInfos_AnimateApp1 = [
  new jit_StaticNodeDebugInfo5([],null,{}),
  new jit_StaticNodeDebugInfo5([],null,{})
]
;
function _View_AnimateApp1(viewUtils,parentInjector,declarationEl) {
  var self = this;
  jit_DebugAppView8.call(this, _View_AnimateApp1,renderType_AnimateApp,jit_ChangeDetectionStrategy_CheckAlways10,viewUtils,parentInjector,declarationEl,jit_ChangeDetectionStrategy_CheckAlways10,nodeDebugInfos_AnimateApp1);
}
_View_AnimateApp1.prototype = Object.create(jit_DebugAppView8.prototype);
_View_AnimateApp1.prototype.createInternal = function(rootSelector) {
  var self = this;
  self._el_0 = self.renderer.createElement(null,'div',self.debug(0,4,0));
  self.renderer.setElementAttribute(self._el_0,'class','box');
  self._text_1 = self.renderer.createText(self._el_0,'',self.debug(1,4,66));
  self._expr_0 = jit_uninitialized0;
  self._expr_1 = jit_uninitialized0;
  self.init([].concat([self._el_0]),[
    self._el_0,
    self._text_1
  ]
  ,[],[]);
  return null;
};
_View_AnimateApp1.prototype.detectChangesInternal = function(throwOnChange) {
  var self = this;
  self.detectContentChildrenChanges(throwOnChange);
  self.debug(0,4,44);
  var currVal_0 = self.parent.context.state;
  if (jit_checkBinding14(throwOnChange,self._expr_0,currVal_0)) {
    var playerResult = AnimateApp_boxAnimation_0(self,self._el_0,self._expr_0,currVal_0);
    self._expr_0 = currVal_0;
  }
  self.debug(1,4,66);
  var currVal_1 = jit_interpolate17(1,'\n  ',self.context.$implicit,'\n');
  if (jit_checkBinding14(throwOnChange,self._expr_1,currVal_1)) {
    self.renderer.setText(self._text_1,currVal_1);
    self._expr_1 = currVal_1;
  }
  self.detectViewChildrenChanges(throwOnChange);
};
_View_AnimateApp1.prototype.detachInternal = function() {
  var self = this;
  var playerResult = AnimateApp_boxAnimation_0(self,self._el_0,self._expr_0,jit_uninitialized0);
};
function viewFactory_AnimateApp1(viewUtils,parentInjector,declarationEl) {
  return new _View_AnimateApp1(viewUtils,parentInjector,declarationEl);
}
return viewFactory_AnimateApp0
})
```

**Comment by tbosch** [comment]
Codgen:

- declare the maps outside of the animation factory function (with component name as prefix)

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+    if (context.styleState == _AnimationStyleStepState.FirstStateChangeStartingStyles) {
+      stylesArr.push(_ANIMATION_START_STATE_STYLES_VAR);
+    }
+
+    ast.styles.forEach(entry => {
+      stylesArr.push(o.literalMap(StringMapWrapper.keys(entry).map(key => [key, o.literal(entry[key])])));
+    });
+
+    return o.importExpr(Identifiers.AnimationStyles).instantiate([
+      _ANIMATION_STYLE_UTIL.callMethod('collectAndResolveStyles', [
```
Comment (tbosch) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: Define an own `CompileIdentifier` with name: `AnimationStyleUtil.collectAndResolveStyles` .

```
new CompileIdentifier({
  name: 'AnimationStyleUtil.collectAndResolveStyles',
  moduleUrl: ...
  runtime: AnimationStyleUtil.collectAndResolveStyles
})
```

File: modules/@angular/core/src/animation/animation_style_util.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
+import {isPresent, isArray} from '../facade/lang';
+import {ListWrapper, StringMapWrapper} from '../facade/collection';
+import {AUTO_STYLE} from './metadata';
+import {FILL_STYLE_FLAG} from '../animation/animation_constants';
```
Comment (tbosch) [inline modules/@angular/core/src/animation/animation_style_util.ts]: this import is wrong, same package.

File: modules/@angular/core/src/animation/animation_style_util.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+  static collectAndResolveStyles(collection: {[key: string]: string|number}, newStyles: {[key: string]: string|number}[]) {
+    newStyles.forEach(entry => {
+      StringMapWrapper.forEach(entry, (value, prop) => {
+        if (value == FILL_STYLE_FLAG) {
+          value = collection[prop];
+          if (!isPresent(value)) {
+            value = AUTO_STYLE;
+          }
+          entry[prop] = value;
```
Comment (tbosch) [inline modules/@angular/core/src/animation/animation_style_util.ts]: Don't mutate but return a cloned objects

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+    context.styleState = _AnimationStyleStepState.FinalStateChangeStartingStyles;
+    var startingStylesExpr = ast.startingStyles.visit(this, context);
+
+    context.styleState = _AnimationStyleStepState.FinalStateChangeKeyframeStyles;
+    var keyframeExpressions = ast.keyframes.map(keyframe => {
+      return keyframe.visit(this, context);
+    });
+
+    context.styleState = _AnimationStyleStepState.FinalStateChangeKeyframeStyles;
+    var keyframesExpr = _ANIMATION_STYLE_UTIL.callMethod('balanceKeyframes', [
```
Comment (tbosch) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: Maybe `targetStateKeyframes`

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+
+    var keyframes = ast.keyframes.map(keyframeEntry => keyframeEntry.visit(this, context));
+
+    context.styleState = _AnimationStyleStepState.NormalStyles;
+
+    var keyframesExpr = o.literalArr(keyframes);
+    return this._callAnimateMethod(ast, startingStylesExpr, keyframesExpr);
+  }
+
+  _visitFinalAnimationStep(ast: AnimationStepAst,
```
Comment (tbosch) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: Rename to `_visitFinalStateAnimation`

File: modules/@angular/compiler/src/animation/animation_compiler.ts (hunk 4 above)
Comment (tbosch) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: Maybe search/replace to `start` / `end` for the start/end state

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+
+    // once complete we want to apply the styles on the element
+    // since the destination state's values should persist once
+    // the animation sequence has completed.
+    statements.push(
+      _ANIMATION_PLAYER_VAR.callMethod('onDone', [
+        o.fn([], [
+          _ANIMATION_FACTORY_RENDERER_VAR.callMethod('setElementStyles', [
+            _ANIMATION_FACTORY_ELEMENT_VAR,
+            _ANIMATION_STYLE_UTIL.callMethod('balanceStyles', [
```
Comment (tbosch) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: Probably you don't need this call, just pass in the final styles

File: modules/@angular/core/src/linker/view.ts [hunk 7]
```diff
@@ -84,6 +93,13 @@ export abstract class AppView<T> {
     }
   }
 
+  registerActiveAnimation(player: AnimationPlayer): void {
```
Comment (tbosch) [inline modules/@angular/core/src/linker/view.ts]: Make this call `player.play` already.
Then rename to `startAnimation`

File: modules/@angular/compiler/src/animation/animation_compiler.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: Or maybe just make these top level functions

**Comment by tbosch** [comment]
Remove special case for `*->*`: Still need to check state name as we could come from a `void`

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+
+      var visitor = new _AnimationBuilder(entry.name, factoryName);
+      compiledAnimations.push(visitor.build(result.ast));
+    });
+    return compiledAnimations;
+  }
+}
+
+enum _AnimationStyleStepState {
+  FirstStateChangeStartingStyles,
```
Comment (tbosch) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: you only need the first one, change to a boolean.

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+    ]);
+  }
+
+  visitAnimationStep(ast: AnimationStepAst, context: _AnimationBuilderContext): o.Expression {
+    if (!isPresent(context.finalAnimateStepAst) && ast.keyframes.length > 0) {
+      context.finalAnimateStepAst = ast;
+      context.styleState = _AnimationStyleStepState.FirstStateChangeStartingStyles;
+    }
+
+    if (context.finalAnimateStepAst === ast && ast.keyframes[0].styles.styles.length == 1) {
```
Comment (tbosch) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: Createa helper function to check this condition, e.g. `emptyAnimationStyles`

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+      let astStyles = ast.keyframes[0].styles.styles[0];
+      if (StringMapWrapper.isEmpty(astStyles)) {
+        return this._visitFinalAnimationStep(ast, context);
+      }
+    }
+
+    var startingStylesExpr = ast.startingStyles.visit(this, context);
+
+    var totalKeyframes = ast.keyframes.length;
+    if (totalKeyframes > 2) {
```
Comment (tbosch) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: This check can go away

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+      keyframesExpr,
+      o.literal(ast.duration),
+      o.literal(ast.delay),
+      o.literal(ast.easing)
+    ]);
+  }
+
+  visitAnimationSequence(ast: AnimationSequenceAst,
+                         context: _AnimationBuilderContext): o.Expression {
+    if (!isPresent(context.finalAnimateStepAst)) {
```
Comment (tbosch) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: Change parser to always have a sequence at the top. So the calculation of `finalAnimateStep` and `firstAnimateStep` should become simpler. Then you can determine these flags at the top level loop (build)

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+        _ANIMATION_FACTORY_ELEMENT_VAR,
+        _ANIMATION_STYLE_UTIL.callMethod('clearStyles', [_ANIMATION_START_STATE_STYLES_VAR]),
+      ]).toStmt());
+
+    ListWrapper.addAll(statements, transitionStatements);
+
+    // in the situation where a any to any state transition (* => *) has
+    // not occurred then we should create a NoOp player so that the onDone
+    // callback can be used for tracking
+    if (!context.fullAnyStateDetected) {
```
Comment (tbosch) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: Remove this optimization

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+  visitAnimationStateTransition(ast: AnimationStateTransitionAst, context: _AnimationBuilderContext): any {
+    context.finalAnimateStepAst = null;
+    var playerExpr = ast.animation.visit(this, context);
+
+    var precondition = _ANIMATION_PLAYER_VAR.equals(o.NULL_EXPR);
+
+    if (ast.fromState == ANY_STATE && ast.toState == ANY_STATE) {
+      context.fullAnyStateDetected = true;
+    } else {
+      if (ast.fromState != ANY_STATE) {
```
Comment (tbosch) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: you always have to have a condition to care for `void`.

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+      _ANIMATION_PLAYER_VAR.set(playerExpr).toStmt()
+    ]);
+  }
+
+  visitAnimationEntry(ast: AnimationEntryAst, context: _AnimationBuilderContext): any {
+    var EMPTY_MAP = o.literalMap([]);
+
+    var statements = [];
+    var transitionStatements = [];
+    ast.definitions.forEach(def => {
```
Comment (tbosch) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: Use 2 arrays in `AnimationEntryAst`: states and transitions

File: modules/@angular/compiler/src/compile_metadata.ts [hunk 15]
```diff
@@ -477,24 +644,30 @@ export class CompileTemplateMetadata {
   templateUrl: string;
   styles: string[];
   styleUrls: string[];
+  animations: CompileAnimationEntryMetadata[];
   ngContentSelectors: string[];
-  constructor({encapsulation, template, templateUrl, styles, styleUrls, ngContentSelectors}: {
+  baseUrl: string;
```
Comment (tbosch) [inline modules/@angular/compiler/src/compile_metadata.ts]: Remove this

File: modules/@angular/compiler/src/compile_metadata.ts [hunk 16]
```diff
@@ -477,24 +644,30 @@ export class CompileTemplateMetadata {
...
+  constructor({encapsulation, template, templateUrl, styles, styleUrls, animations, ngContentSelectors}: {
     encapsulation?: ViewEncapsulation,
     template?: string,
     templateUrl?: string,
     styles?: string[],
     styleUrls?: string[],
-    ngContentSelectors?: string[]
+    ngContentSelectors?: string[],
+    animations?: CompileAnimationEntryMetadata[],
+    baseUrl?: string
```
Comment (tbosch) [inline modules/@angular/compiler/src/compile_metadata.ts]: remove this

File: modules/@angular/compiler/test/animation/animation_compiler_spec.ts [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+      var component = CompileDirectiveMetadata.create({
+        template: new CompileTemplateMetadata({
+          animations: [compiledAnimationEntry]
+        })
+      });
+
+      return compileAnimations(component);
+    };
+
+    xit('should throw an exception containing all the inner animation parser errors', () => {
```
Comment (tbosch) [inline modules/@angular/compiler/test/animation/animation_compiler_spec.ts]: Remove `xit`

File: modules/@angular/core/test/animation/animation_integration_spec.ts [hunk 18]
```diff
@@ -0,0 +1,547 @@
...
+                                         expect(driver.log.length).toEqual(1);
+
+                                         var keyframes2 = driver.log[0]['keyframeLookup'];
+                                         expect(keyframes2.length).toEqual(2);
+                                         expect(keyframes2[0]).toEqual([0, {'opacity': 0}]);
+                                         expect(keyframes2[1]).toEqual([100, {'opacity': 1}]);
+                                       });
+                    })));
+
+                    /*
```
Comment (tbosch) [inline modules/@angular/core/test/animation/animation_integration_spec.ts]: Uncomment or remove

File: modules/@angular/core/test/animation/animation_integration_spec.ts [hunk 19]
```diff
@@ -0,0 +1,547 @@
...
+                       cmp.exp = true;
+                       fixture.detectChanges();
+
+                       flushMicrotasks();
+                       zone.simulateMicrotaskEmpty();
+                     });
+               })));
+      });
+
+      /*
```
Comment (tbosch) [inline modules/@angular/core/test/animation/animation_integration_spec.ts]: Uncomment or remove

File: modules/@angular/compiler/src/animation/animation_parser.ts [hunk 20]
```diff
@@ -0,0 +1,507 @@
...
+      ast.keyframes = [startKeyframe, endKeyframe];
+    }
+  } else if (ast instanceof AnimationWithStepsAst) {
+    ast.steps.forEach(entry => _fillAnimationAstStartingKeyframes(entry, collectedStyles, errors));
+  }
+}
+
+function _parseTimeExpression(exp: string | number,
+                              errors: AnimationParseError[]): _AnimationTimings {
+  var regex = /^([\.\d]+)(m?s)(?:\s+([\.\d]+)(m?s))?(?:\s+([-a-z]+))?/gi;
```
Comment (teropa) [inline modules/@angular/compiler/src/animation/animation_parser.ts]: This does not parse easing functions like `cubic-bezier(.17,.67,.83,.67)`. It only allows `-` or `a-z` in the easing value.

Comment (matsko) [inline modules/@angular/compiler/src/animation/animation_parser.ts]: Korjattu

**Pull request closed without merging**